File: Components/search_reset_creep.m
function [output1,output2] = search_reset_creep(varargin)
% Add a small value (positive or negative) to each element of solutions
% with a probability, for discrete problems with ordinal attributes.

%------------------------------Copyright-----------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 

% Please reference the paper below if using AutoOptLib in your publication:
% @article{zhao2023autooptlib,
%  title={AutoOptLib: A Library of Automatically Designing Metaheuristic 
%         Optimization Algorithms in Matlab},
%  author={Zhao, Qi and Yan, Bai and Hu, Taiwei and Chen, Xianglong and 
%          Yang, Jian and Shi, Yuhui},
%  journal={arXiv preprint 	arXiv:2303.06536},
%  year={2023}
% }
%--------------------------------------------------------------------------

mode = varargin{end};
switch mode
    case 'execute'
        Solution = varargin{1};
        Problem  = varargin{2};
        Para     = varargin{3};
        Aux      = varargin{4};

        Prob = Para(1);
        Amp  = Para(2);
        Amp  = max(1,round((Problem.bound(2,:)-Problem.bound(1,:)+1).*Amp)); % 1*D
        if ~isnumeric(Solution)
            New = Solution.decs;
        else
            New = Solution;
        end
        [N,D] = size(New);

        StepSize = zeros(N,D);
        for i = 1:D
            StepSize(:,i) = randi(Amp(i),N,1); % sample step size for each dimension, N*D
        end
        ind = rand(N,D) < 0.5; % N*D logical indices
        StepSize(ind) = -StepSize(ind);
        Temp = New+StepSize;
        Lower = repmat(Problem.bound(1,:),N,1);
        Upper = repmat(Problem.bound(2,:),N,1);
        Temp = max(min(Temp,Upper),Lower);

        ind = rand(N,D) < Prob; % N*D
        New(ind) = Temp(ind);
        output1 = New;
        output2 = Aux;

    case 'parameter'
        output1 = [0,0.5;0,0.5]; % probability and amplitude 

    case 'behavior'
        output1 = {'LS','small','small';'GS','large','large'}; % small probabilities and amplitudes perform local search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/search_reset_creep.py
"""Creep mutation for discrete ordinal variables."""
from __future__ import annotations

from typing import Any

import numpy as np

from ._utils import ensure_rng, flex_get


def _extract_matrix(solution: Any) -> np.ndarray:
    decs = flex_get(solution, "decs")
    if callable(decs):
        arr = np.asarray(decs(), dtype=int)
    elif decs is not None:
        arr = np.asarray(decs, dtype=int)
    else:
        arr = np.asarray(solution, dtype=int)
    if arr.ndim != 2:
        raise ValueError("Solution must be 2-D for search_reset_creep")
    return arr


def search_reset_creep(*args):
    mode = args[-1]
    if mode == "execute":
        solution = args[0]
        problem = args[1]
        para = args[2] if len(args) > 2 else None
        aux = args[3] if len(args) > 3 else None
        rng = ensure_rng(aux)

        params = np.asarray(para).reshape(-1) if para is not None else np.array([0.1, 0.1])
        prob = float(params[0]) if params.size > 0 else 0.1
        amp_ratio = float(params[1]) if params.size > 1 else 0.1

        new = _extract_matrix(solution)
        bound = np.asarray(flex_get(problem, "bound"), dtype=int)
        lower = bound[0]
        upper = bound[1]
        n, d = new.shape

        amp = np.maximum(1, np.round((upper - lower + 1) * amp_ratio).astype(int))
        step = np.zeros((n, d), dtype=int)
        for j in range(d):
            step[:, j] = rng.integers(1, amp[j] + 1, size=n)
        sign = rng.random((n, d)) < 0.5
        step[sign] *= -1
        candidate = new + step
        candidate = np.maximum(candidate, lower)
        candidate = np.minimum(candidate, upper)

        mask = rng.random((n, d)) < prob
        new[mask] = candidate[mask]
        return new, aux

    if mode == "parameter":
        return [0, 0.5, 0, 0.5], None

    if mode == "behavior":
        return [["LS", "small", "small"], ["GS", "large", "large"]], None

    raise ValueError(f"Unsupported mode: {mode}")